File: backend/app/api/v1/departments.py
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List
from uuid import UUID
from sqlalchemy.ext.asyncio import AsyncSession

from ...database.session import get_db_session
from ...schemas.department import Department, DepartmentDetail, DepartmentCreate, DepartmentUpdate
from ...schemas.common import BaseResponse
from ...services.department_service import DepartmentService
from ...security import AuthContext, get_auth_context
from ...core.exceptions import NotFoundError, PermissionDeniedError, ConflictError, ValidationError, BadRequestError

router = APIRouter(prefix="/departments", tags=["departments"])


@router.get("/", response_model=List[Department])
async def get_departments(
    session: AsyncSession = Depends(get_db_session),
    _: AuthContext = Depends(get_auth_context)  # Ensure user is authenticated
):
    """
    Get all departments for dropdown/selection purposes.
    
    Returns simple department list accessible to all authenticated users.
    For advanced department management with user details, use the specific department endpoint.
    """
    try:
        service = DepartmentService(session)
        result = await service.get_departments_for_dropdown()
        return result
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error fetching departments: {str(e)}"
        )


@router.get("/{department_id}", response_model=DepartmentDetail)
async def get_department(
    department_id: UUID,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """
    Get detailed department information by ID.
    
    Access rules:
    - **Admin/Manager/Viewer**: Can access any department
    - **Supervisor/Employee**: Can only access their own department
    
    Returns enriched department data including:
    - Basic department information
    - User count and manager details
    - List of users in the department
    """
    try:
        service = DepartmentService(session)
        result = await service.get_department_by_id(department_id, context)
        return result
        
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except PermissionDeniedError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=str(e)
        )
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error"
        )


@router.post("/", response_model=Department, status_code=status.HTTP_201_CREATED)
async def create_department(
    department_create: DepartmentCreate,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """
    Create a new department (admin only).
    
    Validates:
    - Department name uniqueness
    - Name length (2-100 characters)
    - User has DEPARTMENT_MANAGE permission
    """
    try:
        service = DepartmentService(session)
        result = await service.create_department(department_create, context)
        return result
        
    except PermissionDeniedError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=str(e)
        )
    except ConflictError as e:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=str(e)
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=str(e)
        )
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error"
        )


@router.put("/{department_id}", response_model=Department)
async def update_department(
    department_id: UUID,
    department_update: DepartmentUpdate,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """
    Update an existing department.
    
    Access rules:
    - **Supervisor and above**: Can update departments
    
    Validates:
    - Department exists
    - Name uniqueness (if changing name)
    - User has appropriate permissions
    """
    try:
        service = DepartmentService(session)
        result = await service.update_department(department_id, department_update, context)
        return result
        
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except PermissionDeniedError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=str(e)
        )
    except ConflictError as e:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=str(e)
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=str(e)
        )
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error"
        )


@router.delete("/{department_id}", response_model=BaseResponse)
async def delete_department(
    department_id: UUID,
    context: AuthContext = Depends(get_auth_context),
    session: AsyncSession = Depends(get_db_session)
):
    """
    Delete a department (admin only).
    
    Business rules:
    - Cannot delete departments with active users
    - Only admin users can delete departments
    - Validates referential integrity before deletion
    
    Returns success message upon successful deletion.
    """
    try:
        service = DepartmentService(session)
        result = await service.delete_department(department_id, context)
        
        # Service returns a dict with message, convert to BaseResponse
        if isinstance(result, dict) and "message" in result:
            return BaseResponse(message=result["message"])
        else:
            return BaseResponse(message="Department deleted successfully")
        
    except NotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )
    except PermissionDeniedError as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=str(e)
        )
    except BadRequestError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except ValidationError as e:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=str(e)
        )
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error"
        )

File: frontend/src/api/server-actions/departments.ts
'use server';

import { revalidateTag } from 'next/cache';
import { departmentsApi } from '../endpoints/departments';
import { createFullyCachedAction, CACHE_TAGS } from '../utils/cache';
import type { 
  Department, 
  DepartmentDetail, 
  DepartmentCreate, 
  DepartmentUpdate,
  UUID,
} from '../types';

/**
 * Server action to get all departments with caching
 */
async function _getDepartmentsAction(): Promise<{
  success: boolean;
  data?: Department[];
  error?: string;
}> {
  try {
    const response = await departmentsApi.getDepartments();
    
    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to fetch departments',
      };
    }
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Get departments action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while fetching departments',
    };
  }
}

export const getDepartmentsAction = createFullyCachedAction(
  _getDepartmentsAction,
  'getDepartments',
  CACHE_TAGS.DEPARTMENTS
);

/**
 * Server action to get a specific department by ID with caching
 */
async function _getDepartmentByIdAction(departmentId: UUID): Promise<{
  success: boolean;
  data?: DepartmentDetail;
  error?: string;
}> {
  try {
    const response = await departmentsApi.getDepartmentById(departmentId);
    
    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to fetch department',
      };
    }
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Get department by ID action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while fetching department',
    };
  }
}

export const getDepartmentByIdAction = createFullyCachedAction(
  _getDepartmentByIdAction,
  'getDepartmentById',
  CACHE_TAGS.DEPARTMENTS
);

/**
 * Server action to create a new department with cache revalidation
 */
export async function createDepartmentAction(departmentData: DepartmentCreate): Promise<{
  success: boolean;
  data?: DepartmentDetail;
  error?: string;
}> {
  try {
    const response = await departmentsApi.createDepartment(departmentData);
    
    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to create department',
      };
    }
    
    // Revalidate departments cache after successful creation
    revalidateTag(CACHE_TAGS.DEPARTMENTS);
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Create department action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while creating department',
    };
  }
}

/**
 * Server action to update an existing department with cache revalidation
 */
export async function updateDepartmentAction(departmentId: UUID, updateData: DepartmentUpdate): Promise<{
  success: boolean;
  data?: DepartmentDetail;
  error?: string;
}> {
  try {
    const response = await departmentsApi.updateDepartment(departmentId, updateData);
    
    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to update department',
      };
    }
    
    // Revalidate departments cache after successful update
    revalidateTag(CACHE_TAGS.DEPARTMENTS);
    
    return {
      success: true,
      data: response.data,
    };
  } catch (error) {
    console.error('Update department action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while updating department',
    };
  }
}

/**
 * Server action to delete a department with cache revalidation
 */
export async function deleteDepartmentAction(departmentId: UUID): Promise<{
  success: boolean;
  error?: string;
}> {
  try {
    const response = await departmentsApi.deleteDepartment(departmentId);
    
    if (!response.success) {
      return {
        success: false,
        error: response.errorMessage || 'Failed to delete department',
      };
    }
    
    // Revalidate departments cache after successful deletion
    revalidateTag(CACHE_TAGS.DEPARTMENTS);
    
    return {
      success: true,
    };
  } catch (error) {
    console.error('Delete department action error:', error);
    return {
      success: false,
      error: 'An unexpected error occurred while deleting department',
    };
  }
}